### PR TITLE
[pkg/stanza] Simplify fingerprint updating

### LINKED ISSUE #22936: Duplicated logs when fingerprint_size is set to larger value
### Component(s)

receiver/filelog

## Description

Filelog receiver constantly reads the same log file when `fingerprint_size` is set to value larger than  [size of  scanner default buffer](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/62c39d6ba132ed5377620a53eebe6a265b5f5f3d/pkg/stanza/fileconsumer/scanner.go#LL14C7-L14C21) and size of the log file is lower than `fingerprint_size` but larger than scanner buffer.

## Steps to Reproduce

- Run otelcontribcol with attached configuration and read [example log file](https://gist.github.com/kkujawa-sumo/60babc6790bf229c6380a6c645a936c8)

## Expected Result
Logs are not duplicated

## Actual Result
Logs are duplicated, the same logs are constantly read

### Collector version

commit: `ca50a98fdda4c362d8782a29f6fc5cc27977f37b`

## Environment
Darwin Kernel Version 22.3.0
go version go1.19.6 darwin/amd64

### OpenTelemetry Collector configuration

```yaml
exporters:
  file:
    path: out.log
receivers:
  filelog:
    fingerprint_size: 17408
    start_at: beginning
    include:
    - generated_log.txt
service:
  pipelines:
    logs:
      exporters:
      - file
      receivers:
      - filelog
```

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by djaglowski**
Depends on #31298

Fixes #22936

This PR changes the way readers update their fingerprints.

Currently,  when `reader.ReadToEnd` is called, it creates a scanner and passes itself (the reader) in as the `io.Reader` so that a custom implementation of `Read` will be used by the scanner. Each time the scanner calls `Read`, we try to perform appropriate reasoning about whether the data we've read should be appended to the fingerprint. The problem is that the correct positioning of the bytes buffer is in some rare cases different than the file's "offset", as we track it. See example [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22937#issuecomment-1576523007).

There appear to be two ways to solve this. A "simple" solution is to independently determine the file handle's current offset with a clever use of `Seek`, ([suggested here](https://stackoverflow.com/a/10901436/3511338). Although this does appear to work, it leaves open the possibility that the fingerprint is corrupted because _if the file was truncated_, we may be updating the fingerprint with incorrect information.

The other solution, proposed in this PR, simply has the custom `Read` function set a flag to indicate that the fingerprint _should_ be updated. Then, just before returning from `ReadToEnd`, we create an entirely new fingerprint. This has the advantage of not having to manage any kind of append operations, but also allows the the opportunity to independently check that the fingerprint has not been altered by truncation.

Benchmarks appear to show all three solutions are close in performance.

### CODE REVIEW

**Comment by djaglowski** [comment]
@kkujawa-sumo, I believe I finally have a solution for https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22936. Please take a look if you would like.

File: pkg/stanza/fileconsumer/internal/reader/reader.go [hunk 1]
```diff
@@ -30,18 +30,20 @@ type Metadata struct {
...
-	decoder         *decode.Decoder
-	headerReader    *header.Reader
-	processFunc     emit.Callback
-	emitFunc        emit.Callback
-	deleteAtEOF     bool
+	logger                 *zap.SugaredLogger
+	fileName               string
+	file                   *os.File
+	fingerprintSize        int
+	initialBufferSize      int
```
Comment (astencel-sumo) [inline pkg/stanza/fileconsumer/internal/reader/reader.go:37]: Am I missing something or is the `initialBufferSize` field only used by the fingerprint tests?

It seems the value of this outside of tests is always equal to `scanner.DefaultBufferSize`, is this true?

Comment (djaglowski) [reply pkg/stanza/fileconsumer/internal/reader/reader.go:37]: I think you are correct and typically wouldn't want to have such a field. However, I think we may want to expose this field to user configuration eventually as well. In a scenario where there are many files being watched, each of which tends to contain only short lines, this could save substantial memory. In the meantime, my thinking is that it's necessary to validate behaviors which we do not have control over (buffer size after initialization).

Comment (astencel-sumo) [reply pkg/stanza/fileconsumer/internal/reader/reader.go:37]: > it's necessary to validate behaviors which we do not have control over

I :100:  agree in principle. In this case, it is lost on me why the unit tests need to be run with various values for the initial buffer size, when in real life the initial buffer size is always equal to 16k.

Anyway, I don't think this should be blocking this PR.

**Review by andrzej-stencel — APPROVED** [review]

File: pkg/stanza/fileconsumer/internal/reader/reader.go (hunk 1 above)
Comment (djaglowski) [reply pkg/stanza/fileconsumer/internal/reader/reader.go:37]: > it is lost on me why the unit tests need to be run with various values for the initial buffer size, when in real life the initial buffer size is always equal to 16k.

Maybe there is a better way but I'm just trying to make sure the code is robust because we only have control over the _initial_ buffer size. In real life, the buffer is resized automatically and I don't want to make assumptions about exactly which values it may take. Most of the test cases attempt to create boundary conditions by reading controlled inputs through various small sizes. I think odds of hitting these boundary conditions are much lower if only working against 16k+ buffers.
